pkg/persistence/object_store.go

```diff
@@ -235,6 +235,16 @@ func (s *objectBackupStore) ListBackups() ([]string, error) {
 		// each of those off to get the backup name.
 		backupName := strings.TrimSuffix(strings.TrimPrefix(prefix, s.layout.subdirs["backups"]), "/")
 
+		// if a bucket is versioned, the s3 folder still exists, even if all s3 objects have been deleted.
+		// We should only take s3 folders having a velero-backup.json object.
+		if ok, errObjectExists := s.objectStore.ObjectExists(s.bucket, s.layout.getBackupMetadataKey(backupName)); !ok || errObjectExists != nil {
```

Comment:
By this check, we only know the backup's metadata file doesn't exist. We don't whether the bucket versioning is on.
I think it's possible some other corner cases are also included by this check.

Reply:
I think it doesn't matter whether it's S3 and versioning is enabled or not or any other storage that knows versioning is in place. As soon as velero-backup.json is missing, but the backup folder name exists, it no longer makes sense to check the folder for velero backups.

Reply:
I guess this would be generic enough for other non s3 cloud provider scenarios which may or may not have the same behavior.